### PR TITLE
Announcing Scala.js 0.6.15.

### PR DESCRIPTION
**Pull request by sjrd**

### CODE REVIEW

**Comment by sjrd** [comment]
Here are already the release notes. I still need to update the docs.

@olafurpg What should I tell in the release notes to people who want to use scalafix to migrate to `@JSGlobal`? I think at some point you showed me a repo with an example, but I can't find it anymore.

**Comment by sjrd** [comment]
Documentation done, and I also updated the tutorial. The updated tutorial code/history is available [in the `dev` branch](https://github.com/scala-js/scalajs-tutorial/commits/dev) (currently using 0.6.15-SNAPSHOT).

**Comment by sjrd** [comment]
Artifacts are on Maven, CLIs are published, and Scaladocs are set up.

**Comment by olafurpg** [comment]
I just released scalafix 0.3.2 which contains necessary changes to run the DemandJSGlobal rewrite. The necessary diff for users to run the rewrite is
```diff
--- /dev/null
+++ b/.scalafix.conf
@@ -0,0 +1,3 @@
+rewrites = [DemandJSGlobal]
+imports.organize = false (optional)
+imports.removeUnused = false (optional)
diff --git a/project/build.properties b/project/build.properties
index 817bc38..27e88aa 100644
--- a/project/build.properties
+++ b/project/build.properties
@@ -1 +1 @@
-sbt.version=0.13.9
+sbt.version=0.13.13
diff --git a/project/plugins.sbt b/project/plugins.sbt
new file mode 100644
index 0000000..442d537
--- /dev/null
+++ b/project/plugins.sbt
@@ -0,0 +1,2 @@
+addSbtPlugin("ch.epfl.scala" % "sbt-scalafix" % "0.3.2")
```
Then `sbt scalafix` will do the rest. I decided to merge the rewrite into the scalafix repo just because it serves as a nice test. However, you can copy [this file](https://github.com/scalacenter/scalafix/blob/f41abea82a0a17b92773dbe0299eadf58d4b04e9/core/src/main/scala/scalafix/rewrite/ScalaJsRewrites.scala) to some scala-js repo and users could run that rewrite by putting into `.scalafix.conf`
```conf
rewrites = ["https://rawgithub...."]
```
More examples on custom rewrites is here: https://scalacenter.github.io/scalafix/#Customrewrites

For troubleshooting, refer users to https://scalacenter.github.io/scalafix/#FAQ/Troubleshooting 

Congrats on the release!

**Comment by sjrd** [comment]
@olafurpg Awesome! I've tested it with scalajs-dom, and it works beautifully! See the PR https://github.com/scala-js/scala-js-dom/pull/277

I'll integrate the setup instructions in the release notes, and announce in the afternoon :)

File: doc/project/building.md [hunk 1]
```diff
@@ -19,22 +19,52 @@ To produce a proper JavaScript file from your code, you need to call the linker:
...
 If your application or one of its libraries requires a DOM (which can be specified with `jsDependencies += RuntimeDOM`), you will also need to install [`jsdom`](https://github.com/tmpvar/jsdom) with `npm install jsdom`.
 
 There are alternative JavaScript interpreters that are available.
 See [JavaScript environments](./js-environments.html) for more details.
 
-### Disabling the optimizations
+### Deprecated: Run without `scalaJSUseMainModuleInitializer`
+
+It is still possible to `run` a Scala.js application that does not have `scalaJSUseMainModuleInitializer := true`.
+However, this is not recommended anymore.
```
Comment (gzm0) [inline doc/project/building.md]: Do we really want to remove this? I realize this is a necessary consequence of getting rid of `@JSExportDescendentObjects` but it seems like a useful feature :S

File: _posts/news/2017-03-21-announcing-scalajs-0.6.15.md [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+tags: [releases]
+permalink: /news/2017/03/21/announcing-scalajs-0.6.15/
+---
+
+
+We are excited to announce the release of Scala.js 0.6.15!
+
+This release was focused on preparations for Scala.js 1.0.0.
+In particular, it should contain the last deprecations for things that are being replaced by better alternatives.
+If your code and build compile without warnings with 0.6.15, it should continue to work unchanged in 1.0.0.
```
Comment (sjrd) [inline _posts/news/2017-03-21-announcing-scalajs-0.6.15.md]: The `jsDependencies` will move to a separate sbt plugin where we can afford to keep crappy things forever ^^

File: _posts/news/2017-03-21-announcing-scalajs-0.6.15.md [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+{% highlight scala %}
+@js.native
+object Foo extends js.Object
+
+@js.native
+@JSName("Foobar")
+class Bar extends js.Object
+{% endhighlight %}
+
+This is now deprecated: `@js.native` objects and classes should now explicitly use the annotation `@JSGlobal` (as opposed to `@JSImport`) as follows:
```
Comment (sjrd) [inline _posts/news/2017-03-21-announcing-scalajs-0.6.15.md]: Hum ... I really meant `JSImport`, in the sense that either you have `@JSGlobal` or `@JSImport`. I guess the parentheses are confusing. I'll remove that.

File: _posts/news/2017-03-21-announcing-scalajs-0.6.15.md [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+{% endhighlight %}
+
+This is redundant.
+You can safely remove the two `@JSExport`s while leaving the behavior unchanged:
+
+{% highlight scala %}
+object Foo extends js.JSApp {
+  def main(): Unit = { ... }
+}
+{% endhighlight %}
```
Comment (sjrd) [inline _posts/news/2017-03-21-announcing-scalajs-0.6.15.md]: `JSApp` still has `@JSExportDescendentObjects`, so the change is safe regardless in 0.6.x. We'll have to decide what to do with `JSApp` with 1.0.0, though, but that's another issue. My current idea is to:
* re-introduce `main(Array[String])` detection in sbt
* use a `ModuleInitializer` that can call those methods
* add `def main(args: Array[String]): Unit = main()` in `js.JSApp` for compatibility
* deprecate `JSApp`

File: _posts/news/2017-03-21-announcing-scalajs-0.6.15.md [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+### `@JSExportDescendentClasses` and `@JSExportDescendentObjects` are deprecated without (direct) replacement
+
+Those two annotations were basically only used by testing frameworks, to "reflectively" instantiate test classes and objects.
+For this use case (and similar ones), read further below about the introduction of the Reflective Instantiation API.
+
+#### Migration tips
+
+**Are you maintaining a testing framework?**
+If you use our `TestUtils.newInstance` and/or `TestUtils.loadModule` methods to perform reflective instantiation of `@JSExportDescendent...` things, all you need to do replace `@JSExportDescendentClasses` and/or `@JSExportDescendentObjects` by `@scala.scalajs.reflect.annotation.EnableReflectiveInstantion`.
```
Comment (sjrd) [inline _posts/news/2017-03-21-announcing-scalajs-0.6.15.md]: Oops, indeed!

File: doc/project/building.md (hunk 1 above)
Comment (sjrd) [inline doc/project/building.md]: I'm not sure it's actually useful.

I'm not certain we'll not have to reinstate that feature in some way, but for now I figured it doesn't hurt to tell people they shouldn't use it. If we don't, we'll never get any complaints, and we won't know that there are use cases for the feature.

**Comment by sjrd** [comment]
Updated.

@gzm0 I have notably been more careful with the "your code will keep working" sentence ;)

File: _posts/news/2017-03-21-announcing-scalajs-0.6.15.md (hunk 4 above)
Comment (gzm0) [inline _posts/news/2017-03-21-announcing-scalajs-0.6.15.md]: SG

**Comment by gzm0** [comment]
:tada:

**Pull request merged**